cmd/extensions/oidc-policy/internal/controller/oidcpolicy_reconciler.go

```diff
@@ -0,0 +1,680 @@
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	gatewayapiv1 "sigs.k8s.io/gateway-api/apis/v1"
+	gatewayapiv1alpha2 "sigs.k8s.io/gateway-api/apis/v1alpha2"
+
+	kuadrantv1 "github.com/kuadrant/kuadrant-operator/api/v1"
+	kuadrantv1alpha1 "github.com/kuadrant/kuadrant-operator/api/v1alpha1"
+	"github.com/kuadrant/kuadrant-operator/internal/reconcilers"
+	extcontroller "github.com/kuadrant/kuadrant-operator/pkg/extension/controller"
+	"github.com/kuadrant/kuadrant-operator/pkg/extension/types"
+	"github.com/kuadrant/kuadrant-operator/pkg/extension/utils"
+)
+
+type OIDCPolicyReconciler struct {
+	*reconcilers.BaseReconciler
+	logger logr.Logger
+}
+
+type ingressGatewayInfo struct {
+	Hostname  string                    `json:"hostname"`
+	Name      string                    `json:"name"`
+	Namespace string                    `json:"namespace"`
+	Protocol  gatewayapiv1.ProtocolType `json:"protocol"`
+	url       *url.URL
+}
+
+func (g *ingressGatewayInfo) GetURL() *url.URL {
+	if g.url == nil {
+		g.url = &url.URL{
+			Scheme: strings.ToLower(string(g.Protocol)),
+			Host:   g.Hostname,
+		}
+	}
+	return g.url
+}
+
+func NewOIDCPolicyReconciler() *OIDCPolicyReconciler {
+	return &OIDCPolicyReconciler{
+		BaseReconciler: reconcilers.NewLazyBaseReconciler(),
+	}
+}
+
+func (r *OIDCPolicyReconciler) WithLogger(logger logr.Logger) *OIDCPolicyReconciler {
+	r.logger = logger
+	return r
+}
+
+// kuadrant permissions
+//+kubebuilder:rbac:groups=kuadrant.io,resources=oidcpolicies,verbs=get;list;watch;update;patch
+//+kubebuilder:rbac:groups=kuadrant.io,resources=oidcpolicies/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=kuadrant.io,resources=oidcpolicies/finalizers,verbs=update
+//+kubebuilder:rbac:groups=kuadrant.io,resources=authpolicies,verbs=get;create;list;watch;update;patch
+//+kubebuilder:rbac:groups=kuadrant.io,resources=authpolicies/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=kuadrant.io,resources=authpolicies/finalizers,verbs=update
+
+// gateway-api permissions
+//+kubebuilder:rbac:groups=gateway.networking.k8s.io,resources=httproutes,verbs=get;create;list;watch;update;patch
+//+kubebuilder:rbac:groups=gateway.networking.k8s.io,resources=httproutes/status,verbs=get;update;patch
+
+// TODO: Current OIDC Workflow works only for Browser apps and Native apps that manage the Auth via browser
+// TODO: It only implements Authentication using the Authorization Code Flow (Recommended). Missing Implicit and Hybrid Flow
+// TODO: Expand TokenSource to work with credentials other than cookie
+
+func (r *OIDCPolicyReconciler) Reconcile(ctx context.Context, request reconcile.Request, kCtx types.KuadrantCtx) (reconcile.Result, error) {
+	r.WithLogger(utils.LoggerFromContext(ctx).WithName("OIDCPolicyReconciler"))
+	r.logger.Info("Reconciling OIDCPolicy")
+
+	oidcPolicy := &kuadrantv1alpha1.OIDCPolicy{}
+	if err := r.Client().Get(ctx, request.NamespacedName, oidcPolicy); err != nil {
```

Comment:
Are we back to reading directly from the cluster in the middle of a reconciliation call? It feels like breaking the state of the world principle.

Reply:
Problem is the pattern. There's no guarantee this Get will return the generation of the resource as it was when reconciliation was triggered. More importantly, a second Get further ahead could introduce yet another state, while decisions have been possibly made already in the same reconciliation call. This is the main motivation of SOTW and a problem the [Policy Machinery](https://github.com/Kuadrant/policy-machinery) offers tools to solve. That's why, IMO, new policy controllers should adopt that pattern, not this one.